Add support for tabs.sendMessage to Web Extension content scripts.

xeenon opened the pull request:
#### cf1260f203c173044feb28d2f481501ac8d313ee
<pre>
Add support for tabs.sendMessage to Web Extension content scripts.
<a href="https://webkit.org/b/261842">https://webkit.org/b/261842</a>
rdar://problem/115805311

Reviewed by Brian Weinstein.

* This adds support for browser.tabs.sendMessage, and browser.runtime.onMessage.
* Cleans up JSON parsing/encoding with some helper functions.
* Introduces WebExtensionFrameIdentifier with conversion to and from WebCore::FrameIdentifier.

* Source/WebKit/DerivedSources-input.xcfilelist: Added WebExtensionMessageSenderParameters.serialization.in.
* Source/WebKit/DerivedSources.make: Ditto.
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::parseJSON): Added.
(WebKit::encodeJSONString): Added.
(WebKit::encodeJSONData): Added.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers): Added WebExtensionMessageSenderParameters.
* Source/WebKit/Shared/Extensions/WebExtensionFrameIdentifier.h: Added.
(WebKit::isMainFrame):
(WebKit::toWebCoreFrameIdentifier):
(WebKit::matchesFrame):
(WebKit::toWebExtensionFrameIdentifier):
(WebKit::toWebAPI):
* Source/WebKit/Shared/Extensions/WebExtensionMessageSenderParameters.h: Copied from Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl.
* Source/WebKit/Shared/Extensions/WebExtensionMessageSenderParameters.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h:
(WebKit::toWebAPI): Added const.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::toWebAPI): Added.
* Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h:
(WebKit::toWebAPI): Added const.
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization _localizationDictionaryForWebExtension:withLocale:]): Drive-by cleanup.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsSendMessage):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension): Use JSON helper.
(WebKit::WebExtension::parseManifest): Ditto.
(WebKit::WebExtension::serializeManifest): Ditto.
(WebKit::WebExtension::serializeLocalization): Ditto.
(WebKit::WebExtension::resourceDataForPath): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::processes const): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::pageListensForEvent const): Added.
(WebKit::WebExtensionContext::processes const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::sendToProcesses): Added.
(WebKit::WebExtensionContext::sendToProcessesForEvent): Use sendToProcesses().
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::onMessage): Added.
(WebKit::WebExtensionContextProxy::dispatchRuntimeMessageEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchRuntimeContentScriptMessageEvent): Added.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::parseSendMessageOptions): Added.
(WebKit::WebExtensionAPITabs::sendMessage): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
(WebKit::toWebPage): Made a RefPtr to match toWebFrame.
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile): Added using WTF. Remove a const that was giving a build error.
(_returnExpression): Added .get() in a few places since a RefPtr is used now.
(_dynamicAttributesImplementation): Ditto.
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::webExtensionContextProxies): Moved here due to ARC issues with the JSON helpers.
(WebKit::WebExtensionContextProxy::get): Ditto.
(WebKit::WebExtensionContextProxy::WebExtensionContextProxy): Ditto.
(WebKit::WebExtensionContextProxy::~WebExtensionContextProxy): Ditto.
(WebKit::WebExtensionContextProxy::getOrCreate): Ditto.
(WebKit::WebExtensionContextProxy::parseLocalization): Ditto.
(WebKit::WebExtensionContextProxy::parseJSON): Deleted.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::enumerateFramesAndNamespaceObjects):
(WebKit::webExtensionContextProxies): Deleted. Moved to WebExtensionContextProxyCocoa.mm.
(WebKit::WebExtensionContextProxy::get): Deleted. Ditto.
(WebKit::WebExtensionContextProxy::getOrCreate): Deleted. Ditto.
(WebKit::WebExtensionContextProxy::WebExtensionContextProxy): Deleted. Ditto.
(WebKit::WebExtensionContextProxy::~WebExtensionContextProxy): Deleted. Ditto.
(WebKit::WebExtensionContextProxy::enumerateNamespaceObjects): Deleted.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Added new tabs.sendMessage tests.

Canonical link: <a href="https://commits.webkit.org/268270@main">https://commits.webkit.org/268270@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a06da48d0a12a2ff7db16237d0ed4b9c5b6baaa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19199 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/19614 "Build was cancelled. Recent messages:OS: Ventura (13.4.1), Xcode: 14.3") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/20212 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21088 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17968 "Failed to checkout and rebase branch from PR 17990") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19414 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22885 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19741 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19420 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19494 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16701 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21961 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/19374 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/16688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17487 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17742 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17662 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21797 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17372 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21732 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2343 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
```diff
@@ -422,6 +422,28 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
+    }
+
+    ASSERT(processes.size() == 1);
+    auto process = processes.takeAny();
```
does this just return any element in the set (so in this case, the only)?

xeenon replied:
Yep

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -111,6 +116,79 @@
+            }
+
+            JSValue *value = dynamic_objc_cast<JSValue>(returnValue);
+            if (!value._isThenable)
```
is `_isThenable` something that returns a promise?

xeenon replied:
Sort of, it is a [JS term](https://masteringjs.io/tutorials/fundamentals/thenable).

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -428,6 +431,28 @@
+    if (!validateDictionary(options, sourceKey, nil, types, outExceptionString))
+        return false;
+
+    if (NSNumber *frameNumber = objectForKey<NSNumber>(options, frameIdKey)) {
```
is it necessary to use `objectForKey` here if it passed the NSNumber.class validation check? i had considered using `objectForKey` as well in my scripting PR but thought it may be unnecessary.

xeenon replied:
Yeah you are right, after validation we can assume the types here. I can change this. I had the same thought when I reviewed your PR.

brentfulgham commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -188,7 +190,42 @@
+NSString *encodeJSONString(NSDictionary *dictionary, NSError **error)
+{
+    if (auto *data = encodeJSONData(dictionary, error))
+        return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
```
This is likely to encourage memory leaks, since we are returning a +1 Ref'd NSString. In typical WebKit code we would return a RetainPtr<NSString> from this method.

brentfulgham replied:
At the very least, this method should probably be called `createEncodedJSONString` to warn the caller that the returned value is +1 Retained.

xeenon replied:
This file is ARC, it is safe to do.

ddkilzer replied:
I believe this is safe (won't leak) because when the method is compiled under ARC (`CocoaHelpers.mm` is compiled with `-fobjc-arc`), ARC will use the naming convention (and lack of `NS_RETURNS_RETAINED` attribute) to autorelease the returned object regardless of whether the caller is ARC or non-ARC.

b-weinstein commented on Source/WebKit/Shared/Extensions/WebExtensionFrameIdentifier.h:
```diff
@@ -0,0 +1,120 @@
+
+static constexpr double MainFrame { 0 };
+
+static constexpr const WebExtensionFrameIdentifier MainFrameIdentifier { std::numeric_limits<uint64_t>::max() - 1 };
```
What's the difference between these two identifiers for the main frame?

xeenon replied:
One is the web API version (`0`) as a `double`, the other is a `WebExtensionFrameIdentifier` object. The identifier classes treat `0` as invalid, so we need to pick another number for this to work, so `std::numeric_limits<uint64_t>::max() - 1`. And we convert between the two.

brentfulgham commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
```diff
@@ -337,6 +343,22 @@ JSObjectRef toJSError(JSContextRef context, NSString *callingAPIName, NSString *
+    if (parameters.url.isValid())
+        result[urlKey] = (NSURL *)parameters.url;
+
+    return [result copy];
```
Why is this returning a copy? Is that to make sure this is +1 Retained on return? It seems like this should either return a RetainPtr<>, or at least be named something like `copyToWebAPI` or `toWebAPICopy` or something to warn the caller that this is returning a potential leak.

xeenon replied:
Also ARC. The copy is to remove the mutability of the dictionary.

ddkilzer replied:
A `-copy` of an `NSMutableDictionary` is also low-cost because Foundation implements copy-on-write semantics when making a non-mutable copy of a mutable dictionary.

brentfulgham commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -111,6 +116,79 @@
+        if (JSValue *value = dynamic_objc_cast<JSValue>(replyMessage))
+            replyMessageJSON = value._toJSONString;
+        else
+            replyMessageJSON = encodeJSONString(replyMessage);
```
In this branch, `replyMessageJSON` is now +1 Retained and will leak.

xeenon replied:
Also an ARC file.

brentfulgham commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
```diff
@@ -337,6 +343,22 @@ JSObjectRef toJSError(JSContextRef context, NSString *callingAPIName, NSString *
     return locale.languageCode;
 }
 
+NSDictionary *toWebAPI(const WebExtensionMessageSenderParameters& parameters)
```
Note that the implementation of `toErrorString` leaks memory. The `formattedUnderlyingErrorString` variable is allocated, but never released. Sometimes it is returned as a +1 Retained object, and other times it is discarded and an autoreleased value is returned.

xeenon replied:
This is ARC as well.

ddkilzer replied:
This won't leak for the same reason as `encodeJSONString()`.  The easy way to check that a file is compiled with ARC in WebKit is to do this:

```
$ grep 'WebExtensionUtilities.mm in Sources' Source/WebKit/WebKit.xcodeproj/project.pbxproj 
		337822482947FBA5002106BB /* WebExtensionUtilities.mm in Sources */ = {isa = PBXBuildFile; fileRef = 337822462947FBA4002106BB /* WebExtensionUtilities.mm */; settings = {COMPILER_FLAGS = "-fobjc-arc"; }; };
				337822482947FBA5002106BB /* WebExtensionUtilities.mm in Sources */,
```

brentfulgham commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -428,6 +431,28 @@
+    if (NSNumber *frameNumber = objectForKey<NSNumber>(options, frameIdKey)) {
+        auto identifier = toWebExtensionFrameIdentifier(frameNumber.doubleValue);
+        if (!identifier) {
+            *outExceptionString = toErrorString(nil, frameIdKey, @"it is not a frame identifier");
```
Sometimes `outExceptionString` is +1 Retained, sometimes it is 0 Retained, which will either lead to a leak or to an over-release.

xeenon replied:
This is ARC as well.

brentfulgham commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -838,6 +863,44 @@ static bool isValid(std::optional<WebExtensionTabIdentifier> identifier, NSStrin
+        return;
+
+    if (message.length > webExtensionMaxMessageLength) {
+        *outExceptionString = toErrorString(nil, @"message", @"it exceeded the maximum allowed length");
```
Ditto the leak or overrelease here.

xeenon replied:
This is ARC as well.

brentfulgham commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
```diff
@@ -27,37 +27,74 @@
 }
 
-RetainPtr<_WKWebExtensionLocalization> WebExtensionContextProxy::parseLocalization(API::Data& json)
+_WKWebExtensionLocalization *WebExtensionContextProxy::parseLocalization(API::Data& json)
```
I don't think we should stop wrapping this in a RetainPtr, since it is allocated inside this method.

xeenon replied:
The file is ARC, so the RetainPtr wasn't needed.

brentfulgham reviewed (changes requested):
I think this change introduces a number of Cocoa-related Retain bugs. @ddkilzer: Could you double-check me on this?

xeenon committed:
Add support for tabs.sendMessage to Web Extension content scripts.
https://webkit.org/b/261842
rdar://problem/115805311

Reviewed by Brian Weinstein.

* This adds support for browser.tabs.sendMessage, and browser.runtime.onMessage.
* Cleans up JSON parsing/encoding with some helper functions.
* Introduces WebExtensionFrameIdentifier with conversion to and from WebCore::FrameIdentifier.

* Source/WebKit/DerivedSources-input.xcfilelist: Added WebExtensionMessageSenderParameters.serialization.in.
* Source/WebKit/DerivedSources.make: Ditto.
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::parseJSON): Added.
(WebKit::encodeJSONString): Added.
(WebKit::encodeJSONData): Added.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers): Added WebExtensionMessageSenderParameters.
* Source/WebKit/Shared/Extensions/WebExtensionFrameIdentifier.h: Added.
(WebKit::isMainFrame):
(WebKit::toWebCoreFrameIdentifier):
(WebKit::matchesFrame):
(WebKit::toWebExtensionFrameIdentifier):
(WebKit::toWebAPI):
* Source/WebKit/Shared/Extensions/WebExtensionMessageSenderParameters.h: Copied from Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl.
* Source/WebKit/Shared/Extensions/WebExtensionMessageSenderParameters.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h:
(WebKit::toWebAPI): Added const.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::toWebAPI): Added.
* Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h:
(WebKit::toWebAPI): Added const.
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization _localizationDictionaryForWebExtension:withLocale:]): Drive-by cleanup.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsSendMessage):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension): Use JSON helper.
(WebKit::WebExtension::parseManifest): Ditto.
(WebKit::WebExtension::serializeManifest): Ditto.
(WebKit::WebExtension::serializeLocalization): Ditto.
(WebKit::WebExtension::resourceDataForPath): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::processes const): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::pageListensForEvent const): Added.
(WebKit::WebExtensionContext::processes const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::sendToProcesses): Added.
(WebKit::WebExtensionContext::sendToProcessesForEvent): Use sendToProcesses().
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::onMessage): Added.
(WebKit::WebExtensionContextProxy::dispatchRuntimeMessageEvent): Added.
(WebKit::WebExtensionContextProxy::dispatchRuntimeContentScriptMessageEvent): Added.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::parseSendMessageOptions): Added.
(WebKit::WebExtensionAPITabs::sendMessage): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
(WebKit::toWebPage): Made a RefPtr to match toWebFrame.
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile): Added using WTF. Remove a const that was giving a build error.
(_returnExpression): Added .get() in a few places since a RefPtr is used now.
(_dynamicAttributesImplementation): Ditto.
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::webExtensionContextProxies): Moved here due to ARC issues with the JSON helpers.
(WebKit::WebExtensionContextProxy::get): Ditto.
(WebKit::WebExtensionContextProxy::WebExtensionContextProxy): Ditto.
(WebKit::WebExtensionContextProxy::~WebExtensionContextProxy): Ditto.
(WebKit::WebExtensionContextProxy::getOrCreate): Ditto.
(WebKit::WebExtensionContextProxy::parseLocalization): Ditto.
(WebKit::WebExtensionContextProxy::parseJSON): Deleted.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::enumerateFramesAndNamespaceObjects):
(WebKit::webExtensionContextProxies): Deleted. Moved to WebExtensionContextProxyCocoa.mm.
(WebKit::WebExtensionContextProxy::get): Deleted. Ditto.
(WebKit::WebExtensionContextProxy::getOrCreate): Deleted. Ditto.
(WebKit::WebExtensionContextProxy::WebExtensionContextProxy): Deleted. Ditto.
(WebKit::WebExtensionContextProxy::~WebExtensionContextProxy): Deleted. Ditto.
(WebKit::WebExtensionContextProxy::enumerateNamespaceObjects): Deleted.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Added new tabs.sendMessage tests.

Canonical link: https://commits.webkit.org/268270@main
